core/src/main/java/org/zenoss/zep/index/impl/BaseEventIndexBackend.java

```diff
@@ -34,12 +38,44 @@
 
     private final Map<String, SS> savedSearches = new ConcurrentHashMap<String, SS>();
 
+    private final LoadingCache<EventFilter, EventTagSeveritiesSet> tagSeveritiesCache;
+
+    private static final Date IMMEDIATELY = new Date(0);
     private static final Logger logger = LoggerFactory.getLogger(EventIndexBackend.class);
 
-    protected BaseEventIndexBackend(Messages messages, TaskScheduler scheduler, UUIDGenerator uuidGenerator) {
+    protected BaseEventIndexBackend(Messages messages, TaskScheduler scheduler, UUIDGenerator uuidGenerator,
+                                    int tagSeverityCacheSize, int tagSeveritiesCacheTTL) {
         this.messages = messages;
         this.scheduler = scheduler;
         this.uuidGenerator = uuidGenerator;
+        this.tagSeveritiesCache = CacheBuilder.
+                newBuilder().
+                maximumSize(tagSeverityCacheSize).
+                refreshAfterWrite(1, TimeUnit.SECONDS).
```

Comment:
Suppose the cache contains entries for foo, bar, and baz. When getting foo, the result will be returned immediately, and foo's entry will be refreshed asynchronously (Note: a refresh of bar and baz is not scheduled). This is no more load than previously. In fact, it's possibly less load -- the cache takes care that concurrent refreshes of the same entry are unified -- only one will actually do the computation.